### PR TITLE
Adopt itanium C++-style mangling for CPU and CUDA targets

### PR DESCRIPTION
**Pull request by sklam**
* Update and fixes to the existing itanium mangler
* Change CPU and CUDA target to use the mangler
* Add `.mangling_args()` method for types to specialize their name
    * e.g. Array will now show like a C++ template class

The use of itanium mangling scheme will shorten most symbol name and allow debuggers/profilers to show a more human-readable name.

### CODE REVIEW

File: numba/itanium_mangler.py [hunk 1]
```diff
@@ -2,15 +2,48 @@
 Itanium CXX ABI Mangler
 
 Reference: http://mentorembedded.github.io/cxx-abi/abi.html
+
+The basic of the mangling scheme.
```
Comment (stuartarchibald) [inline numba/itanium_mangler.py]: s/basic/basics/

File: numba/itanium_mangler.py [hunk 2]
```diff
@@ -2,15 +2,48 @@
...
+
+We are hijacking the CXX mangling scheme for our use.  We map Python modules
+into CXX namespace.  A `module1.submodule2.foo` is mapped to
+`module1::submodule2::foo`.   For parameterized numba types, we treat them as
+templated types; for example, `array(int64, 1d, C)` becomes an
+`array<int64, 1, C>`.
+
+All mangled names are prefixed with "_Z".  It is followed by the name of the
+entity.  A name contains one or more identifiers.  Each identifier is encoded
+in as "<num of char><name>".   If the name is namespaced and, therefore,
```
Comment (stuartarchibald) [inline numba/itanium_mangler.py]: s/in//

File: numba/itanium_mangler.py [hunk 3]
```diff
@@ -2,15 +2,48 @@
...
+multiple identifiers, the entire name is encoded as "N<name>E".
```
Comment (stuartarchibald) [inline numba/itanium_mangler.py]: I think this needs `has multiple identifiers`

File: numba/itanium_mangler.py [hunk 4]
```diff
@@ -2,15 +2,48 @@
...
+
+For functions, arguments types follow.  There are condensed encoding for basic
```
Comment (stuartarchibald) [inline numba/itanium_mangler.py]: s/encoding/encodings/

File: numba/itanium_mangler.py [hunk 5]
```diff
@@ -2,15 +2,48 @@
...
+scheme for them to avoid leading digits.
 """
 
 from __future__ import print_function, absolute_import
-from numba import types
+
 import re
 
+from numba import types
+from numba import types, utils
```
Comment (stuartarchibald) [inline numba/itanium_mangler.py]: `types` appears twice

File: numba/utils.py [hunk 6]
```diff
@@ -33,6 +33,7 @@
     get_ident = threading.get_ident
     intern = sys.intern
     file_replace = os.replace
+    asbyteint = int
```
Comment (stuartarchibald) [inline numba/utils.py]: Don't think this is needed?

Comment (sklam) [inline numba/utils.py]: I am using `int` as an identity function.

**Pull request merged**